src/hl/container.rs

```diff
@@ -346,6 +347,115 @@ impl<'a> Writer<'a> {
     }
 }
 
+#[derive(Debug)]
+pub struct ByteReader<'a> {
+    obj: &'a Container,
+    pos: u64,
+    dt: Datatype,
+    obj_space: Dataspace,
+    xfer: PropertyList,
+}
+
+impl<'a> ByteReader<'a> {
+    pub fn new(obj: &'a Container) -> Result<Self> {
+        ensure!(!obj.is_attr(), "ByteReader cannot be used on attribute datasets");
+
+        let file_dtype = obj.dtype()?;
+        let mem_dtype = Datatype::from_type::<u8>()?;
+        file_dtype.ensure_convertible(&mem_dtype, Conversion::NoOp)?;
+
+        let obj_space = obj.space()?;
+        ensure!(obj_space.shape().len() == 1, "Only rank 1 datasets can be read vie ByteReader.");
```

Comment:
Do we really care if it's rank 1 or rank 2? What's stopping from reading rank N dataset?

Reply:
My intuition was that a reader using ByteReader will have no information about the shape and stride. Therefore, it doesn't really make sense. Or do you mean using a ByteReader on a slice?

Reply:
`Seek` would be quite odd for a higher dimensioned dataset, and would be really odd if the dataset was resized. We could remove this restriction in the future if we decide on semantics for such iterations